gnucash/import-export/qif-imp/qif-to-gnc.scm

```diff
@@ -449,6 +455,88 @@
           (qif-file:xtns qif-file)))
        sorted-qif-files-list)
 
+      ;; Add prices
+      (let ((pricedb (gnc-pricedb-get-db (gnc-get-current-book)))
+            (symbol-hash (make-symbol-hash stock-map))
+            (sorted-prices '())
+            (pruned-prices '())
+            (pruned-price-count 0)
+            (commodity #f)
+            (current-symbol " invalid symbol "))
+
+        ;; Let's combine all the prices and sort them; it
+        ;;  is slightly more efficient to do them by symbol and
+        ;;  tidies up error reporting on missing symbols
+        (for-each
+          (lambda (qif-file)
+            (set! sorted-prices (append sorted-prices (qif-file:prices qif-file))))
+         qif-files-list)
+
+        (set! sorted-prices (sort sorted-prices
+          (lambda (a b)
+            (string-ci<? (qif-price:symbol a) (qif-price:symbol b)))))
+
+        (gnc-pricedb-begin-edit pricedb)
+        ;; Turning off bulk update avoids an n^2 performance
+        ;;  cost, but also turns off duplicate checking
+        (gnc-pricedb-set-bulk-update pricedb #t)
+
+        (if progress-dialog
+            (gnc-progress-dialog-set-sub progress-dialog
+                (string-append (G_ "Discarding duplicate prices"))))
+
+        ;; Prune duplicate records before inserting
+        (for-each
+          (lambda (price)
+            (if (not (string-ci=? current-symbol (qif-price:symbol price)))
+                (begin
+                  (set! current-symbol (qif-price:symbol price))
+                  (set! commodity (hash-ref symbol-hash current-symbol))
+
+                  (if (equal? commodity #f)
+                      (qif-import:log progress-dialog
+                          "qif-import:qif-to-gnc"
+                          (G_ (format #f "Warning: cannot find commodity for symbol ~a." current-symbol))))))
+            (if (not (equal? commodity #f))
+                (if (not (duplicate-price? pricedb commodity default-currency price))
+                    (set! pruned-prices (cons price pruned-prices))
+                    (set! pruned-price-count (+ 1 pruned-price-count))))
+            (update-progress))
+         sorted-prices)
+        (set! pruned-prices (reverse pruned-prices))
```

Comment:
A few comments if you want: you're concatenating several lists into 1 long list via `append`, then the long list is sorted, and filtered. 

1. it's marginally faster to use `append-reverse` instead of `append` -- see https://github.com/cky/guile/blob/89ce9fb31b00f1f243fe6f2450db50372cc0b86d/libguile/srfi-1.c#L81. 

2. IMHO instead of `append-reverse` it's better to filter and accumulate at the same time, then sort at the end.

Would you want to squeeze a few milliseconds from this?

Reply:
Thanks for the tip on append.

My instinct is that it's not worth the optimization.  The large file I'm using takes over 15 minutes to import, without the prices; the prices bit adds perhaps a minute to the overall timeline.  A few milliseconds won't make much difference to the user experience.   I also suspect that the vast majority of import operations are done against exactly one qif file.

I did initially write this code without the sorting, and it sorted + filtered as it went and was dissatisfied and rewrote it into this form. You get some efficiency from sorting, as you don't need to look up the commodity for each transaction that way.  It also makes for a natural way of emitting a warning when a symbol is not found only once.  The original code generated one warning for every price record with a missing symbol, which polluted the log.

Reply:
It would be useful (for us) to understand your large file. e.g. number of accounts, transactions, prices including invalid prices, etc. A sample large datafile with bogus but realistic data _would_ be nice. IMO 15mins wait is 14min too long.

Reply:
FWIW what would block this merge is that we can confirm it works well with a sample (small or large) datafile with transactions and prices. In bugzilla would be ok too.

Reply:
Let me think on that.  It's 30 years of personal history and I have an obvious concern regarding security.   I've got some scripts that transform QIF files; I'll see if I can make myself comfortable that I can sanitize it.

I ran perf against it, and my initial results didn't give me any sense of aha!  The bulk of time (86%) is in the trans_cleanup_commit call stack which does not feel startling to me.

If I trust quiffen, it's got 54 accounts, 26 securities, and 115 categories.  There are 23K transactions, with the bulk of them in a single checking account.  The qif file itself is about 5.4M; about half of that is used for the ~70k prices records.  It's a typical qif file with those flaws, and Quicken had been allowed to 'archive' some data, which screws up the precision of the data in some frustrating ways.

Note that I have found a few other flaws with the import which I'm hoping to figure out and ideally submit patches for.  For example, a transfer of a security from one fund to another seems to fail to result in a shares only transaction; instead half is shares, half is cash.  I'd like to see if I can 'undo' the damage of the Quicken archive; I see a path to do it manually, and I'm unsure if it could be automated.  I'm also hoping to agitate for an auto prefixing of account names, but I need to do more homework first.

At the very least, I'll pull together some representative sample files that illustrate what I've seen.

Reply:
@christopherlam : I checked with my wife, and she is uncomfortable with me releasing that data, no matter how sanitized I think it may be.  I'm willing to break out samples, and to run tests against it if that will help.  I did add a sample file that shows prices to this PR.

Reply:
Thanks for adding this datafile, which does include an invalid price entry which is nicely handled:
```
Line 83: Could not parse price line: "ABC",," 1/18'38"
```

Is it intentional that it displays warnings about missing symbols DEF and GHI when importing into a blank datafile?
```
Warning: cannot find commodity for symbol DEF.
Warning: cannot find commodity for symbol GHI.
```

Reply:
Yes, those are also intentional errors.